evm/estimators/gas.py

```diff
@@ -0,0 +1,85 @@
+from cytoolz import (
+    curry,
+)
+
+
+from evm.exceptions import (
+    OutOfGas,
+)
+from evm.utils.spoof import (
+    SpoofTransaction,
+)
+
+
+@curry
+def execute_plus_buffer(multiplier, state, transaction):
+    computation = state.execute_transaction(transaction)
+
+    if computation.is_error:
+        raise computation._error
+
+    gas_used = transaction.gas_used_by(computation)
+
+    gas_plus_buffer = int(gas_used * multiplier)
+
+    return min(gas_plus_buffer, state.gas_limit)
+
+
+double_execution_cost = execute_plus_buffer(2)
+
+
+def _is_enough_gas(state, transaction):
+    snapshot = state.snapshot()
+    computation = state.execute_transaction(transaction)
+    state.revert(snapshot)
+
+    if computation.is_error:
+        return not isinstance(computation._error, OutOfGas)
```

Comment:
I'm not sure that we should only be checking for `OutOfGas` here.  I got annoyed that pyethereum only had one exception type (which is accurate in terms of VM semantics).  Hence, py-evm has fine grained exceptions for all the various ways that errors can blow up your VM execution.  However, as far as the EVM is concerned, every single `VMError` subclass is an out-of-gas error.

I think it's more appropriate to check against the base `VMError`.  I can think of at least one contract in the ethereum alarm clock contracts which would throw an `InvalidJumpDestination` error in py-evm if the gas allotment for the transaction was too low.

Reply:
> I'm not sure that we should only be checking for `OutOfGas` here

In theory, I liked the idea that a throw or other kind of error would not cause this estimate to choose the block maximum.

>  I can think of at least one contract in the ethereum alarm clock contracts which would throw an InvalidJumpDestination error in py-evm if the gas allotment for the transaction was too low.

... which ruins the whole plan above. I think the best solution is to throw if there is an error at the block limit (as you mentioned in another comment), and let the caller decide what's appropriate to do. Then all later errors can be considered out-of-gas related (even if not `OutOfGas`), since the transaction has succeeded once, at the block limit.